keps/sig-storage/1234-20190129-csi-migration.md

```diff
@@ -0,0 +1,518 @@
+
+There is a push to remove CloudProvider code from kubernetes.
+
+There will not be any general auto-deployment mechanism for ALL CSI drivers
+covered in this document so the timeline to remove CloudProvider code using this
+design is undetermined: For example: At some point GKE could auto-deploy the GCE
+PD CSI driver and have migration for that turned on by default, however it may
+not deploy any other drivers by default. And at this point we can only remove
+the code for the GCE In-tree plugin (this would still break anyone doing their
+own deployments while using GCE unless they install the GCE PD CSI Driver).
+
+We could have auto-deploy depending on what cloud provider kubernetes is running
+on. But AFAIK there is no standard mechanism to guarantee this on all Cloud
+Providers.
+
+For example the requirements for just the GCE Cloud Provider code for storage
+with minimal disruption to users would be:
+* In-tree to CSI Plugin migration goes GA
+* GCE PD CSI Driver deployed on GCE/GKE by default (resource requirements of
+  driver need to be determined)
+* GCE PD CSI Migration turned on by default
+* Remove in-tree plugin code and cloud provider code
+
+And at this point users doing their own deployment and not installing the GCE PD
+CSI driver encounter an error.
+
+### Testing
+
+#### Standard
+
+Good news is that all “normal functionality” can be tested by simply bringing up
+a cluster with “migrated” drivers and running the existing e2e tests for that
+driver. We will create CI jobs that run in this configuration for each new
+volume plugin
+
+#### Migration/Non-migration (Upgrade/Downgrade)
+
+Write tests were in a normal workflow of attach/mount/unmount/detach, we have
+any one of these operations actually happen with the old volume plugin, not the
+CSI one This makes sure that the workflow is resiliant to rollback at any point
+in time.
+
+#### Version Skew
+
+Master/Node can have up to 2 version skw. Master must always be equal or higher
+version than the node. It should be covered by the tests in
+Migration/Non-migration section.
+
+#### Scalability & Performance
+
+Migration adds some overhead in the form of API calls, a new RPC boundary,
+driver latency, and shim latency. It is important that we document the current
+performance of in-tree drivers then test and document those drivers with
+migration. The performance hit must be within acceptable ranges TBD. We define
+"volume lifecycle operations" as including
+attach/detach/mount/unmount/mountdevice/unmountdevice/resize/etc. Important
+scalability & performance metrics include but are not limited to:
+* Supported number of concurrent volume lifecycle operations
+* Time to complete volume lifecycle operations
+  * 50th percentile, 99th percentile, etc.
+
+
+### Risks and Mitigations
+
+* Performance risks as outlined above
+* ADC and Kubelet syncronization fairly complicated, upgrade path non-trivial - mitigation discussed above
+
+## Graduation Criteria
```

Comment:
I think for this section we want to detail the criteria for moving from alpha->beta and beta->ga.  Things like what kinds of tests do we need to have, when are things turned on by default, etc